### PR TITLE
Add finalizer

### PR DESCRIPTION
**Pull request by legionus**

### CODE REVIEW

File: pkg/operator/bootstrap.go [hunk 1]
```diff
@@ -67,9 +75,12 @@ func (h *Handler) Bootstrap() (*regopapi.ImageRegistry, error) {
...
 					VolumeSource: corev1.VolumeSource{
-						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
-							ClaimName: "image-registry-pvc",
-						},
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+						/*
+							PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+								ClaimName: "image-registry-pvc",
+							},
+						*/
```
Comment (bparees) [inline pkg/operator/bootstrap.go]: why this change?

File: pkg/operator/finalizer.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+				},
+			}
+
+			err := sdk.Get(cr)
+			if err != nil {
+				return fmt.Errorf("failed to get custom resource %s/%s: %s", o.ObjectMeta.Namespace, o.ObjectMeta.Name, err)
+			}
+		}
+
+		cr.ObjectMeta.Finalizers = finalizers
```
Comment (bparees) [inline pkg/operator/finalizer.go]: I don't think you can do this.  What if the update that caused the conflict was someone else adding a finalizer?  Now you'll have removed their finalizer.

File: pkg/operator/finalizer.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		kerrors.IsTimeout,
+		kerrors.IsServerTimeout,
+		kerrors.IsTooManyRequests,
+		utilnet.IsProbableEOF,
+		utilnet.IsConnectionReset,
+	}
+
+	retryTime := 3 * time.Second
+
+	err = wait.PollInfinite(retryTime, func() (stop bool, err error) {
```
Comment (bparees) [inline pkg/operator/finalizer.go]: why do we need to wait for the deletion to happen?

File: pkg/operator/bootstrap.go (hunk 1 above)
Comment (legionus) [reply pkg/operator/bootstrap.go]: Ops. my mistake.

File: pkg/operator/finalizer.go (hunk 2 above)
Comment (legionus) [reply pkg/operator/finalizer.go]: Hm. How I should remove my finalizer ?

Comment (bparees) [reply pkg/operator/finalizer.go]: i think you need to iterate the list again and remove it again, since the list may have changed.

Comment (bparees) [reply pkg/operator/finalizer.go]: (that is, the same thing you're doing on line 59)

File: pkg/operator/finalizer.go (hunk 3 above)
Comment (legionus) [reply pkg/operator/finalizer.go]: We do not allow two resources in the same space.

Comment (bparees) [reply pkg/operator/finalizer.go]: hm.  So you are waiting here until it goes away before you (re)create the default samplesresource?

Comment (bparees) [reply pkg/operator/finalizer.go]: err, image registry resource.

File: pkg/operator/finalizer.go (hunk 2 above)
Comment (bparees) [reply pkg/operator/finalizer.go]: in practice? probably no one will.  But in theory, anyone w/ sufficient permission could and i'd rather not leave a potential hole open in case we do later introduce another finalizer for some reason and had forgotten about this.

Comment (legionus) [reply pkg/operator/finalizer.go]: Why then do we not follow the situation specifying our operator as an initializer (in theory this is also possible)?

File: pkg/operator/finalizer.go (hunk 3 above)
Comment (legionus) [reply pkg/operator/finalizer.go]: > So you are waiting here until it goes away before you (re)create the default samplesresource?

I'm waiting for the removal of this particular resource so that you can create another resource and not finalize this one again (we can miss a delete event by definition).

File: pkg/operator/finalizer.go (hunk 2 above)
Comment (bparees) [reply pkg/operator/finalizer.go]: What would that accomplish for us?

File: pkg/operator/finalizer.go (hunk 3 above)
Comment (bparees) [reply pkg/operator/finalizer.go]: makes sense, thanks

**Comment by legionus** [comment]
@bparees comments have been addressed.

File: pkg/operator/finalizer.go (hunk 2 above)
Comment (dmage) [reply pkg/operator/finalizer.go]: @legionus they are mainly made for mutating/validating admission controllers.

File: pkg/operator/finalizer.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+		if !kerrors.IsNotFound(err) {
+			for _, isRetryError := range errorFuncs {
+				if isRetryError(err) {
+					return
+				}
+			}
+
+			// If the error sends the Retry-After header, we respect it as an explicit confirmation we should retry.
+			if delaySeconds, shouldRetry := kerrors.SuggestsClientDelay(err); shouldRetry {
+				delayTime := time.Duration(delaySeconds)
```
Comment (dmage) [inline pkg/operator/finalizer.go]: time.Duration represents the time in nanoseconds. You should multiply the result by time.Second.

File: pkg/operator/finalizer.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	err = generate.RemoveConfigState(configState)
+	if err != nil {
+		return fmt.Errorf("unable to remove previous config state: %s", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) finalizeResources(o *regopapi.ImageRegistry) error {
+	defer utilruntime.HandleCrash()
```
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/operator/finalizer.go]: Removed

File: pkg/operator/finalizer.go (hunk 4 above)
Comment (legionus) [reply pkg/operator/finalizer.go]: Ooops. Cut-and-Paste error. Sorry. Fixed.

File: pkg/operator/finalizer.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+)
+
+func (h *Handler) RemoveResources(o *regopapi.ImageRegistry) error {
+	modified := false
+
+	templetes, err := h.GenerateTemplates(o, &h.params)
+	if err != nil {
+		return fmt.Errorf("unable to genetate templates: %s", err)
```
Comment (bparees) [inline pkg/operator/finalizer.go]: typo: generate

Comment (legionus) [reply pkg/operator/finalizer.go]: Fixed
